File: CloudFlow/FrontEnd-App/src/app/shared/service/document-type.service.ts
import { Injectable } from "@angular/core";
import { ApiCaller } from "../api-caller";
import { AsyncCallback } from "../async-callback";
import { DocumentTypeModel } from "../model/content/document-type.model";
import { AppError } from "../model/app-error";
import { ApiPathConstants } from "../constants/api-path.constants";
import { ApiPathUtils } from "./../utils";

@Injectable()
export class DocumentTypeService {

	private apiCaller: ApiCaller;

	public constructor(apiCaller: ApiCaller) {
		this.apiCaller = apiCaller;
	}

	public getAllDocumentTypesForDisplay(callback: AsyncCallback<DocumentTypeModel[], AppError>): void {
		this.apiCaller.call(ApiPathConstants.GET_ALL_DOCUMENT_TYPES_FOR_DISPLAY, null, DocumentTypeModel, callback);
	}

	public getDocumentTypeById(documentTypeId: number, callback: AsyncCallback<DocumentTypeModel, AppError>): void {
		this.apiCaller.call(ApiPathUtils.appendParametersToPath(ApiPathConstants.GET_DOCUMENT_TYPE_BY_ID, "" + documentTypeId), null, DocumentTypeModel, callback);
	}

	public getAvailableDocumentTypes(callback: AsyncCallback<DocumentTypeModel[], AppError>): void {
		this.apiCaller.call(ApiPathConstants.GET_AVAILABLE_DOCUMENT_TYPES, null, DocumentTypeModel, callback);
	}

	public getAvailableDocumentTypesForSearch(callback: AsyncCallback<DocumentTypeModel[], AppError>): void {
		this.apiCaller.call(ApiPathConstants.GET_AVAILABLE_DOCUMENT_TYPES_FOR_SEARCH, null, DocumentTypeModel, callback);
	}

	public deleteDocumentType(documentTypeId: number, callback: AsyncCallback<void, AppError>): void {
		let relativePath: string = ApiPathUtils.appendParametersToPath(ApiPathConstants.DELETE_DOCUMENT_TYPE, documentTypeId.toString());
		this.apiCaller.call(relativePath, null, null, callback);
	}

	public saveDocumentType(documentType: DocumentTypeModel, callback: AsyncCallback<void, AppError>): void {
		this.apiCaller.call(ApiPathConstants.SAVE_DOCUMENT_TYPE, documentType, null, callback);
	}

	public getDocumentTypesWithNoWorkflow(callback: AsyncCallback<DocumentTypeModel[], AppError>): void {
		this.apiCaller.call(ApiPathConstants.GET_DOCUMENT_TYPES_WITH_NO_WORKFLOW, null, DocumentTypeModel, callback);
	}

	public existDocumentTypeWithName(documentTypeName: string, callback: AsyncCallback<boolean, AppError>): void {
		let relativePath: string = ApiPathUtils.appendParametersToPath(ApiPathConstants.EXIST_DOCUMENT_TYPE_WITH_NAME, documentTypeName);
		this.apiCaller.call(relativePath, null, Boolean, callback);
	}
	
	public getDocumentTypeIdByName(documentTypeName: string, callback: AsyncCallback<number, AppError>): void {
		let relativePath: string = ApiPathUtils.appendParametersToPath(ApiPathConstants.GET_DOCUMENT_TYPE_BY_NAME, documentTypeName);
		this.apiCaller.call(relativePath, null, Number, callback);
	}
}

File: CloudFlow/src/ro/cloudSoft/cloudDoc/services/content/DocumentTypeService.java
package ro.cloudSoft.cloudDoc.services.content;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.google.common.collect.SetMultimap;

import ro.cloudSoft.cloudDoc.core.AppException;
import ro.cloudSoft.cloudDoc.domain.content.DocumentCreationInDefaultLocationView;
import ro.cloudSoft.cloudDoc.domain.content.DocumentType;
import ro.cloudSoft.cloudDoc.domain.content.DocumentTypeTemplate;
import ro.cloudSoft.cloudDoc.domain.content.UserMetadataDefinition;
import ro.cloudSoft.cloudDoc.domain.security.SecurityManager;

public interface DocumentTypeService {
	
	List<DocumentType> getAllDocumentTypesForDisplay();
	
	DocumentType getDocumentTypeById(Long id);
	
	DocumentType getDocumentTypeById(Long id, SecurityManager userSecurity);

	/**
	 * Salveaza tipul de document, impreuna cu template-urile sale.
	 * Daca tipul de document era nou, dupa salvare va avea ID-ul completat.
	 * 
	 * @param documentType tipul de document
	 * @param templates template-urile tipului de document
	 * @param namesForTemplatesToDelete numele template-urilor care trebuie sterse
	 * @param userSecurity datele utilizatorului curent
	 * 
	 * @return ID-ul tipului de document
	 */
	Long saveDocumentType(DocumentType documentType, Collection<DocumentTypeTemplate> templates,
		Collection<String> namesForTemplatesToDelete, SecurityManager userSecurity) throws AppException;
	
	void deleteDocumentType(Long id, SecurityManager userSecurity) throws AppException;
	
	List<UserMetadataDefinition> getUserMetadataDefinitions(List<Long> documentTypeIds);

	/**
	 * Returneaza un Map cu etichetele item-ulor posibili din lista, grupate dupa
	 * valorile lor, apoi dupa numele metadatei de tip lista pentru care sunt item-ii.
	 * 
	 * @param documentTypeId ID-ul tipului de document pentru care sa se aduca valorile
	 */
	Map<String, Map<String, String>> getListItemLabelByListItemValueByMetadataName(Long documentTypeId);
	
	List<DocumentType> getAvailableDocumentTypes(SecurityManager userSecurity);
	
	List<DocumentCreationInDefaultLocationView> getDocumentCreationInDefaultLocationViews(SecurityManager userSecurity);
	
	List<DocumentType> getAvailableDocumentTypesForSearch(SecurityManager userSecurity);
	
	Map<Long, String> getDocumentTypesNameMap(Set<Long> ids);
	
	List<DocumentType> getDocumentTypesWithNoWorkflow();
	
	DocumentTypeTemplate getTemplate(Long documentTypeId, String fileName);

	List<DocumentTypeTemplate> getTemplates(Long documentTypeId);
	
	/**
	 * Aduce numele tuturor metadatelor care apartin de tipurile de documente cu ID-urile specificate,
	 * grupate dupa ID-ul tipului de document.
	 */
	SetMultimap<Long, String> getMetadataNamesByDocumentTypeId(Collection<Long> documentTypeIds);

	boolean isMimeTypeUsedInDocumentTypes(Long mimeTypeId);
	
	String getDocumentTypeName(Long id);
	
	boolean existDocumentTypeWithName(String name);
	
	DocumentType getDocumentTypeByName(String documentTypeName);
	
	Long getDocumentTypeIdByName(String documentTypeName);
	
	List<DocumentType> getDocumentTypesByNames(String... documentTypeNames);
	
	Long getMetadataDefinitionIdByNameAndDocumentType(String name, Long documentTypeId);
	
}